File: lib/db_conn.js
/**
 * lib/db_conn.js — Helper único de conexiones a BD (Node)
 * Lee credenciales de `integracion_conexionesbd.env` en la raíz del repo.
 * Provee pools mysql2 para BDs locales y remotas (via SSH tunnel con ssh2).
 *
 * Uso:
 *   const db = require('../../lib/db_conn')
 *   const poolEffi = await db.local('effi_data')
 *   const poolGes  = await db.gestion()
 *   const [rows]   = await poolGes.execute('SELECT 1')
 *
 * Cada BD remota usa su propio puerto local para evitar colisión cuando
 * se abren múltiples tuneles simultáneos (VPS + Hostinger).
 */

const mysql = require('mysql2/promise')
const net   = require('net')
const fs    = require('fs')
const path  = require('path')
// ssh2 se carga lazy — solo módulos que llaman a .integracion()/.gestion()/.comunidad() lo necesitan
let _ssh2 = null
function _loadSSH2() {
  if (!_ssh2) _ssh2 = require('ssh2')
  return _ssh2
}

// ─── Cargar .env central ─────────────────────────────────────────────
const ENV_PATH = path.resolve(__dirname, '..', 'integracion_conexionesbd.env')
function cargarEnv(ruta) {
  if (!fs.existsSync(ruta)) {
    console.warn(`[db_conn] WARN: ${ruta} no existe, usando variables de entorno del sistema`)
    return
  }
  fs.readFileSync(ruta, 'utf8').split('\n').forEach(linea => {
    const l = linea.trim()
    if (!l || l.startsWith('#')) return
    const i = l.indexOf('=')
    if (i < 0) return
    const k = l.slice(0, i).trim()
    const v = l.slice(i + 1).trim()
    if (!(k in process.env)) process.env[k] = v
  })
}
cargarEnv(ENV_PATH)

// TIMEZONE — lee de APP_TIMEZONE (fuente única); DB_TIMEZONE queda como alias legacy.
const TIMEZONE = process.env.APP_TIMEZONE || process.env.DB_TIMEZONE || '-05:00'

// ─── Lectura de config ───────────────────────────────────────────────
function _cfgLocal() {
  return {
    host: process.env.DB_LOCAL_HOST || '127.0.0.1',
    port: parseInt(process.env.DB_LOCAL_PORT || '3306', 10),
    user: process.env.DB_LOCAL_USER,
    password: process.env.DB_LOCAL_PASS,
  }
}

function _cfgRemota(prefijo) {
  const P = prefijo
  return {
    sshHost:    process.env[`DB_${P}_SSH_HOST`],
    sshPort:    parseInt(process.env[`DB_${P}_SSH_PORT`] || '22', 10),
    sshUser:    process.env[`DB_${P}_SSH_USER`],
    sshKey:     process.env[`DB_${P}_SSH_KEY`],
    remoteHost: process.env[`DB_${P}_REMOTE_HOST`] || '127.0.0.1',
    remotePort: parseInt(process.env[`DB_${P}_REMOTE_PORT`] || '3306', 10),
    localPort:  parseInt(process.env[`DB_${P}_LOCAL_PORT`], 10),
    dbName:     process.env[`DB_${P}_NAME`],
    dbUser:     process.env[`DB_${P}_USER`],
    dbPass:     process.env[`DB_${P}_PASS`],
  }
}

// ─── Pools locales (cached por nombre de BD) ─────────────────────────
const _poolsLocales = {}
function poolLocal(dbName) {
  if (_poolsLocales[dbName]) return _poolsLocales[dbName]
  const c = _cfgLocal()
  const pool = mysql.createPool({
    ...c, database: dbName,
    timezone: TIMEZONE, dateStrings: true,
    waitForConnections: true, connectionLimit: 10,
  })
  pool.pool.on('connection', conn => {
    conn.query(`SET time_zone = '${TIMEZONE}'`)
  })
  _poolsLocales[dbName] = pool
  return pool
}

// ─── Conexiones remotas (SSH tunnel + pool por prefijo) ──────────────
const _remotas = {
  INTEGRACION: { sshClient: null, tcpServer: null, ready: false, pool: null, reconectando: false },
  GESTION:     { sshClient: null, tcpServer: null, ready: false, pool: null, reconectando: false },
  INVENTARIO:  { sshClient: null, tcpServer: null, ready: false, pool: null, reconectando: false },
  COMUNIDAD:   { sshClient: null, tcpServer: null, ready: false, pool: null, reconectando: false },
}

function _crearPoolRemoto(prefijo, cfg) {
  const st = _remotas[prefijo]
  st.pool = mysql.createPool({
    host: '127.0.0.1', port: cfg.localPort,
    user: cfg.dbUser, password: cfg.dbPass, database: cfg.dbName,
    timezone: TIMEZONE, dateStrings: true,
    waitForConnections: true, connectionLimit: 10,
  })
  st.pool.pool.on('connection', conn => {
    conn.query(`SET time_zone = '${TIMEZONE}'`)
  })
}

function _destruirPoolRemoto(prefijo) {
  const st = _remotas[prefijo]
  if (st.pool) st.pool.end().catch(() => {})
  st.pool = null
}

function _conectarSSH(prefijo, cfg) {
  const st = _remotas[prefijo]
  return new Promise((resolve, reject) => {
    if (st.sshClient) { try { st.sshClient.destroy() } catch (_) {} st.sshClient = null }
    const key = fs.readFileSync(cfg.sshKey)
    const { Client } = _loadSSH2()
    const client = new Client()
    st.sshClient = client

    client.on('ready', () => {
      console.log(`[db_conn:${prefijo}] SSH tunnel listo → ${cfg.sshUser}@${cfg.sshHost}:${cfg.sshPort}`)
      _crearPoolRemoto(prefijo, cfg)
      st.ready = true
      resolve()
    })
    client.on('error', err => {
      console.error(`[db_conn:${prefijo}] SSH error:`, err.message)
      if (!st.ready) reject(err)
    })
    client.on('close', () => {
      if (st.sshClient === client && st.ready) {
        console.log(`[db_conn:${prefijo}] SSH tunnel cerrado — programando reconexión`)
        st.ready = false
        _destruirPoolRemoto(prefijo)
        st.sshClient = null
        _programarReconexion(prefijo, cfg, 5000)
      }
    })
    client.connect({ host: cfg.sshHost, port: cfg.sshPort, username: cfg.sshUser, privateKey: key })
  })
}

function _crearTcpServer(prefijo, cfg) {
  const st = _remotas[prefijo]
  return new Promise((resolve, reject) => {
    st.tcpServer = net.createServer(localSock => {
      if (!st.sshClient) { localSock.end(); return }
      st.sshClient.forwardOut('127.0.0.1', cfg.localPort, cfg.remoteHost, cfg.remotePort,
        (err, stream) => {
          if (err) { console.error(`[db_conn:${prefijo}] forwardOut error:`, err.message); localSock.end(); return }
          localSock.pipe(stream).pipe(localSock)
          localSock.on('error', () => stream.end())
          stream.on('error', () => localSock.end())
        }
      )
    })
    st.tcpServer.on('error', err => {
      if (err.code === 'EADDRINUSE') {
        console.log(`[db_conn:${prefijo}] puerto ${cfg.localPort} ocupado — asumiendo tunnel activo`)
        _crearPoolRemoto(prefijo, cfg)
        st.ready = true
        return resolve()
      }
      reject(err)
    })
    st.tcpServer.listen(cfg.localPort, '127.0.0.1', () => resolve())
  })
}

function _programarReconexion(prefijo, cfg, delay) {
  const st = _remotas[prefijo]
  if (st.reconectando) return
  st.reconectando = true
  setTimeout(() => {
    _conectarSSH(prefijo, cfg)
      .then(() => { st.reconectando = false })
      .catch(err => {
        console.error(`[db_conn:${prefijo}] reconexión falló:`, err.message)
        st.reconectando = false
        _programarReconexion(prefijo, cfg, 15000)
      })
  }, delay)
}

// Detecta si la BD remota es realmente local (mismo servidor) — salta el tunnel SSH
function _esLocal(cfg) {
  const h = (cfg.sshHost || '').toLowerCase()
  return h === 'localhost' || h === '127.0.0.1' || h === '' || h === 'direct'
}

async function _conectarRemota(prefijo) {
  const st = _remotas[prefijo]
  if (st.ready && st.pool) return st.pool
  const cfg = _cfgRemota(prefijo)
  if (!cfg.dbName) {
    throw new Error(`Config DB_${prefijo}_* incompleta en integracion_conexionesbd.env`)
  }

  // Modo directo (sin SSH): cuando la BD está en el mismo servidor que esta app
  if (_esLocal(cfg)) {
    console.log(`[db_conn:${prefijo}] modo directo (sin SSH tunnel) → ${cfg.remoteHost}:${cfg.remotePort}`)
    st.pool = mysql.createPool({
      host: cfg.remoteHost, port: cfg.remotePort,
      user: cfg.dbUser, password: cfg.dbPass, database: cfg.dbName,
      timezone: TIMEZONE, dateStrings: true,
      waitForConnections: true, connectionLimit: 10,
    })
    st.pool.pool.on('connection', conn => {
      conn.query(`SET time_zone = '${TIMEZONE}'`)
    })
    st.ready = true
    return st.pool
  }

  // Modo SSH tunnel (default)
  if (!cfg.sshHost) {
    throw new Error(`Config DB_${prefijo}_SSH_* incompleta en integracion_conexionesbd.env`)
  }
  if (!st.tcpServer) await _crearTcpServer(prefijo, cfg)
  if (st.ready) return st.pool  // EADDRINUSE path
  await _conectarSSH(prefijo, cfg)
  return st.pool
}

// ─── API pública ─────────────────────────────────────────────────────
// Getters sync (poolGestion etc) leen del state interno EN CADA ACCESO, así
// cuando el SSH se reconecta y el pool se recrea, los consumers ven el pool
// nuevo sin tener que refrescar cachés locales.
module.exports = {
  TIMEZONE,
  // Pool a BD local
  local(dbName) { return poolLocal(dbName) },
  // Async: abren tunnel+pool si aún no listo. Usar en `conectar()` del bootstrap.
  integracion() { return _conectarRemota('INTEGRACION') },
  gestion()     { return _conectarRemota('GESTION') },
  inventario()  { return _conectarRemota('INVENTARIO') },
  comunidad()   { return _conectarRemota('COMUNIDAD') },
  // Sync getters al pool actual (pueden ser null si aún no conectado o cerrando)
  get poolIntegracion() { return _remotas.INTEGRACION.pool },
  get poolGestion()     { return _remotas.GESTION.pool },
  get poolInventario()  { return _remotas.INVENTARIO.pool },
  get poolComunidad()   { return _remotas.COMUNIDAD.pool },
}

File: scripts/lib/db_conn.py
"""
scripts/lib/db_conn.py — Helper único de conexiones a BD (Python)
Lee credenciales de `integracion_conexionesbd.env` en la raíz del repo.

Uso con context manager (recomendado):
    from lib.db_conn import local, integracion, gestion, comunidad

    with local('effi_data') as conn:
        cur = conn.cursor()
        cur.execute('SELECT 1')

    with integracion(dict_cursor=True) as conn:
        ...

Uso con dicts raw (para scripts legados que quieren armar la conexión manual):
    from lib.db_conn import cfg_local, cfg_remota_ssh, cfg_remota_db

Mismos drivers que ya se usan: pymysql, sshtunnel.
"""
import os
import sys
from pathlib import Path
from contextlib import contextmanager

import pymysql
from sshtunnel import SSHTunnelForwarder
from dotenv import load_dotenv

# ─── Cargar .env central (raíz del repo) ─────────────────────────────
_RAIZ_REPO = Path(__file__).resolve().parents[2]
_ENV_PATH  = _RAIZ_REPO / 'integracion_conexionesbd.env'
if _ENV_PATH.exists():
    load_dotenv(_ENV_PATH)
else:
    print(f'[db_conn] WARN: {_ENV_PATH} no existe', file=sys.stderr)

TIMEZONE = os.getenv('DB_TIMEZONE', '-05:00')


# ─── Config readers ──────────────────────────────────────────────────
def cfg_local():
    """Dict con host/port/user/password del MariaDB local (sin database)."""
    return dict(
        host=os.getenv('DB_LOCAL_HOST', '127.0.0.1'),
        port=int(os.getenv('DB_LOCAL_PORT', 3306)),
        user=os.getenv('DB_LOCAL_USER'),
        password=os.getenv('DB_LOCAL_PASS'),
    )


def cfg_remota_ssh(prefijo):
    """Config SSH para una BD remota. prefijo ∈ {INTEGRACION, GESTION, INVENTARIO, COMUNIDAD}."""
    P = prefijo.upper()
    return dict(
        host=os.getenv(f'DB_{P}_SSH_HOST'),
        port=int(os.getenv(f'DB_{P}_SSH_PORT', 22)),
        user=os.getenv(f'DB_{P}_SSH_USER'),
        key=os.path.expanduser(os.getenv(f'DB_{P}_SSH_KEY', '')),
        remote_host=os.getenv(f'DB_{P}_REMOTE_HOST', '127.0.0.1'),
        remote_port=int(os.getenv(f'DB_{P}_REMOTE_PORT', 3306)),
    )


def cfg_remota_db(prefijo):
    """Dict user/password/database listo para pymysql/mysql.connector (sin host/port)."""
    P = prefijo.upper()
    return dict(
        user=os.getenv(f'DB_{P}_USER'),
        password=os.getenv(f'DB_{P}_PASS'),
        database=os.getenv(f'DB_{P}_NAME'),
        charset='utf8mb4',
    )


# ─── Tuneles SSH compartidos (por prefijo) ───────────────────────────
_tuneles = {}  # prefijo → SSHTunnelForwarder activo


def _es_local(prefijo):
    """Detecta si la BD remota es realmente local (mismo servidor). Salta SSH tunnel."""
    h = (os.getenv(f'DB_{prefijo.upper()}_SSH_HOST', '') or '').lower()
    return h in ('localhost', '127.0.0.1', '', 'direct')


def abrir_tunel(prefijo):
    """Abre (o reusa) un SSHTunnelForwarder para el prefijo dado. Retorna el tunel.

    Si DB_<prefijo>_SSH_HOST es localhost/127.0.0.1/direct, retorna None: el caller
    debe conectar directo al MariaDB local sin pasar por tunnel."""
    P = prefijo.upper()
    if _es_local(P):
        return None
    if P in _tuneles and _tuneles[P].is_active:
        return _tuneles[P]
    s = cfg_remota_ssh(P)
    t = SSHTunnelForwarder(
        (s['host'], s['port']),
        ssh_username=s['user'],
        ssh_pkey=s['key'],
        remote_bind_address=(s['remote_host'], s['remote_port']),
    )
    t.start()
    _tuneles[P] = t
    return t


def cerrar_tuneles():
    """Cierra todos los tuneles activos. Útil al final de scripts de una sola pasada."""
    for t in _tuneles.values():
        try:
            t.stop()
        except Exception:
            pass
    _tuneles.clear()


# ─── Context managers (API principal) ────────────────────────────────
def _conectar(kwargs, dict_cursor):
    k = dict(kwargs)
    k.setdefault('charset', 'utf8mb4')
    k.setdefault('autocommit', True)
    if dict_cursor:
        k['cursorclass'] = pymysql.cursors.DictCursor
    return pymysql.connect(init_command=f"SET time_zone = '{TIMEZONE}'", **k)


@contextmanager
def local(db_name, dict_cursor=False):
    """Conexión a una BD del MariaDB local. Uso: with local('effi_data') as conn:"""
    conn = _conectar(dict(cfg_local(), database=db_name), dict_cursor)
    try:
        yield conn
    finally:
        conn.close()


@contextmanager
def remota(prefijo, dict_cursor=False):
    """Conexión a una BD remota. prefijo ∈ {INTEGRACION, GESTION, INVENTARIO, COMUNIDAD}.

    Usa SSH tunnel salvo que DB_<prefijo>_SSH_HOST sea localhost/127.0.0.1/direct,
    en cuyo caso conecta directo al MariaDB del mismo servidor."""
    tunel = abrir_tunel(prefijo)
    db = cfg_remota_db(prefijo)
    ssh = cfg_remota_ssh(prefijo)
    if tunel is None:
        host = ssh['remote_host']
        port = ssh['remote_port']
    else:
        host = '127.0.0.1'
        port = tunel.local_bind_port
    kwargs = dict(
        host=host, port=port,
        user=db['user'], password=db['password'], database=db['database'],
        connect_timeout=15, read_timeout=30,
    )
    conn = _conectar(kwargs, dict_cursor)
    try:
        yield conn
    finally:
        conn.close()


def integracion(dict_cursor=False): return remota('INTEGRACION', dict_cursor)
def gestion(dict_cursor=False):     return remota('GESTION',     dict_cursor)
def inventario(dict_cursor=False):  return remota('INVENTARIO',  dict_cursor)
def comunidad(dict_cursor=False):   return remota('COMUNIDAD',   dict_cursor)


# ─── cfg_inventario(): dict listo para pymysql.connect con tunnel SSH abierto ───
# Útil para scripts heredados que hacen pymysql.connect(**DB_INV) estilo dict.
# Abre el tunnel 1 vez por proceso y reusa.
def cfg_inventario(dict_cursor=True):
    """Retorna dict compatible con pymysql.connect(**).
    Abre tunnel SSH al VPS si no está abierto, conecta al puerto local forwardeado.
    """
    return _cfg_remota_dict('INVENTARIO', dict_cursor)


def cfg_integracion(dict_cursor=True):
    """Retorna dict compatible con pymysql.connect(**) para os_integracion en VPS.
    Abre tunnel SSH al VPS si no está abierto, conecta al puerto local forwardeado.

    USAR PARA: scripts/apps de inventario, producción, gestión, ERP que necesiten
    consultar tablas zeffi_* (OPs, materiales, artículos, recetas, costos) o
    resumen_ventas_*. NUNCA usar cfg_local()+'effi_data' para esto — effi_data
    local es solo intermediaria del pipeline (ver MANIFESTO §8).
    """
    return _cfg_remota_dict('INTEGRACION', dict_cursor)


def _cfg_remota_dict(prefijo, dict_cursor):
    t = abrir_tunel(prefijo)
    db = cfg_remota_db(prefijo)
    cfg = {
        'host': '127.0.0.1',
        'port': t.local_bind_port if t else db['remote_port'],
        'user': db['user'],
        'password': db['password'],
        'database': db['database'],
        'charset': 'utf8mb4',
        'autocommit': True,
        'connect_timeout': 15,
    }
    if dict_cursor:
        cfg['cursorclass'] = pymysql.cursors.DictCursor
    return cfg
